arch/risc-v/include/syscall.h

```diff
@@ -124,16 +124,10 @@
 #define SYS_signal_handler_return (7)
 #endif /* !CONFIG_BUILD_FLAT */
 
-#if defined (CONFIG_ARCH_USE_S_MODE) && defined (__KERNEL__)
-#  define ASM_SYS_CALL \
-     " addi sp, sp, -16\n"                  /* Make room */ \
-     REGSTORE " ra, 0(sp)\n"                /* Save ra */ \
-     " jal  ra, riscv_dispatch_syscall\n"   /* Dispatch (modifies ra) */ \
-     REGLOAD " ra, 0(sp)\n"                 /* Restore ra */ \
-     " addi sp, sp, 16\n"                   /* Restore sp */
+#if defined(CONFIG_ARCH_USE_S_MODE) && defined(__KERNEL__)
+#  define ASM_SYS_CALL "j riscv_dispatch_syscall\n"
```

Comment:
I don't think this will work, up_signal_dispatch actually returns here and now the software will simply get lost. ecall knows where to return since it sets EPC.

Reply:
> I don't think this will work, up_signal_dispatch actually returns here and now the software will simply get lost because no link is done.
> 

it depends on whether there is additional statement after sys_callx: the change will work fine if it's a simple return statement.

> ecall knows where to return since it sets EPC.

Reply:
I think it would be possible to tailcall here *if* the function prototype was not inline. I still think the ra value does not point to the caller of sys_callx, it points to where the function calling the inlined C-procedure would return.

Assembly macro would work, no link is needed for that.

Reply:
Yes, this should now work, but this leaves libc without a proper implementation for sys_callx, does it not ? It's very unfortunate that the userspace uses the same definitions for sys_call() as the kernel does.

Reply:
Another way is used alias to redirect sys_callx to riscv_dispatch_syscall.

Reply:
@pussuw this patch is ready for review now. Could you take a look and try with your kernel mode environment?

Reply:
Yes this should work. The procedure call to syscallx will do the jump-and-link.